### PR TITLE
Extend Core Modules Part 2 - Replace Explicit Module Bean Instantiations

### PR DESCRIPTION
**Pull request by Jason-Dang**
## Description

Replace explicit instantiations of module classes with BeanFactory usage to ensure custom module definitions are respected.

**Replaced 2375 total occurrence(s) of 96 search string(s) in 582 of 7212 files.**

**Relation/dependency/includes  https://github.com/salesagility/SuiteCRM/pull/7795**

### Changes:
- Use script to iterate all files and replace occurrences of explicit instantiations with `BeanFactory::newBean`, prefixing with backslash if file uses namespaces.
- Commit every 10 files and then any remainder at the end.
- Include file ranges in commit headers.
- Detail what was replaced in each file in commit body.

##### Part 3 - Replace Dynamic Instantiations.
   - Occurrences of `new {ModuleClass}()` replace with `\BeanFactory::newBean('{ModuleName}')`
- Replace dynamic instantiations of core modules' classes with Bean Factory usage.
   - Occurrences of `new $moduleClass()` etc. replace with `\BeanFactory::newBean('{ModuleName}')`.
##### Part 4 - Replace Module Class Name used as Table.
- Replace usages of the lower-cased module name as the module table name in some queries.
   - Occurrences where the module name is retrieved from `beanList` key(s), lower-cased and used as the table name, replace with `$bean->getTableName();` (may need to instantiate a bean in some cases or better yet add a static property and method for static info like table names etc.).

**Note:** *For the above the module directory or plural name needs to be used instead of the class/object name such as `Accounts` instead of `Account`.*

## Motivation and Context

Previously unable to extend almost all core module classes without breaking system in various areas. Those that could be extended, don't have their custom classes respected everywhere in the CRM.

## How To Test This

- Run unit tests with `./vendor/bin/phpunit --bootstrap tests/bootstrap.php tests/unit/phpunit/data/BeanFactoryTest.php` (There is a flag in BeanFactoryTestCase to indicate whether to test all modules - default is `false`).
- Smoke test the entire CRM with and without extending core module classes.
- Extend core modules using new custom global arrays (Example below).
 **Custom module classes will only be respected by areas that use the Bean Factory.**

```
$customBeanList['Accounts'] = 'CustomAccount';
$customObjectList['Accounts'] = 'CustomAccount';
$customBeanFiles['Account'] = 'custom/modules/Accounts/CustomAccount.php';
```

OR

```
$customBeanList['Accounts'] = 'CustomAccount';
$customObjectList['Accounts'] = 'CustomAccount';
$customBeanFiles['CustomAccount'] = 'custom/modules/Accounts/CustomAccount.php';
```

*As above, either the core class name or custom class name can be used as the key for the bean file entry.*

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

Stonks Up

### CODE REVIEW

**Comment by Jason-Dang** [comment]
Looks like all the tests got updated and are causing conflicts, instead of manually resolving each file, going to accept the new upstream changes and then re-run the script.

**Comment by connorshea** [comment]
Hmm, how long does it take to run the script? I'd think it'd be easier to just remove all the commits from this PR, rebase on top of hotfix-7.10.x, then re-run the script and force-push, rather than having a second set of commits that makes reviewing this a more difficult.

**Comment by connorshea** [comment]
Also, sorry for making this process complicated 🙈

**Comment by Jason-Dang** [comment]
@connorshea Not long just several minutes :) No worries at all, I should be making it easier to review so it's all stuff to bear in min in the future too :D I've rebased twice lol first time to just before the script run and second time I just fully rebased on hotfix-7.10.x to keep the script stuff to one commit plus 59 from the script one and 1 last commit to remove the script.

Just spinning up my vm now to run the script :)

File: tests/unit/phpunit/SuitePHPUnit_Framework_TestCase.php [hunk 1]
```diff
@@ -52,7 +52,7 @@ public function setUp()
 
         global $current_user, $sugar_config;
         try {
-            $current_user = @\BeanFactory::getBean('Users'); //new User();
+            $current_user = @\BeanFactory::getBean('Users'); //\BeanFactory::newBean('Users');
```
Comment (connorshea) [inline tests/unit/phpunit/SuitePHPUnit_Framework_TestCase.php]: I don't know if we should change the comment here?

File: modules/MailMerge/Merge.php [hunk 2]
```diff
@@ -107,7 +107,7 @@
 
     $fields =  get_field_list($seed);
 
-    $document = new DocumentRevision();//new Document();
+    $document = BeanFactory::newBean('DocumentRevisions');//BeanFactory::newBean('Documents');
```
Comment (connorshea) [inline modules/MailMerge/Merge.php]: We can probably remove this comment?

File: modules/Import/Importer.php [hunk 3]
```diff
@@ -579,7 +579,7 @@ protected function saveMappingFile()
 
         $firstrow    = json_decode(html_entity_decode($_REQUEST['firstrow']), true);
         $mappingValsArr = $this->importColumns;
-        $mapping_file = new ImportMap();
+        $mapping_file = BeanFactory::newBean('Import_1');
```
Comment (connorshea) [inline modules/Import/Importer.php]: What's this?

File: modules/Import/Importer.php [hunk 4]
```diff
@@ -748,7 +748,7 @@ public function setFieldKeyTranslator($translator)
      */
     protected function _undoCreatedBeans(array $ids)
     {
-        $focus = new UsersLastImport();
+        $focus = BeanFactory::newBean('Import_2');
```
Comment (connorshea) [inline modules/Import/Importer.php]: And this?

File: modules/Import/controller.php [hunk 5]
```diff
@@ -106,12 +106,12 @@ public function action_mapping()
         $results = array('message' => '');
         // handle publishing and deleting import maps
         if (isset($_REQUEST['delete_map_id'])) {
-            $import_map = new ImportMap();
+            $import_map = BeanFactory::newBean('Import_1');
             $import_map->mark_deleted($_REQUEST['delete_map_id']);
```
Comment (connorshea) [inline modules/Import/controller.php]: And these, and the others named Import_1 and Import_2

**Comment by Jason-Dang** [comment]
@conorshea Thanks for all the time and effort reviewing! Appreciate it was a big PR with a lot of commits/files.

I’ll have an in-depth look into the stuff you’ve flagged. Off hand I think the comments should be removed and for the import stuff, all classes replaced were taken from the bean list. I’ll need to double check and make sure that these are compatible with the bean factory.

I’ll have to owe you pint for that mind-numbing...

File: tests/unit/phpunit/SuitePHPUnit_Framework_TestCase.php (hunk 1 above)
Comment (Jason-Dang) [reply tests/unit/phpunit/SuitePHPUnit_Framework_TestCase.php]: The comments were changed by the script, my opinion would be that they should be removed but I’ll leave that to the Product Team to decide. I’ll leave them as is for now as I think it reflects the change, if the comment is meant to be there it reflects the same expression but matches its current state.

**Comment by connorshea** [comment]
so, what's holding this one up? :)

**Comment by Mac-Rae** [comment]
@Jason-Dang If you get the chance to resolve these conflics ill get this tested asap (Promise this time :rofl:)

**Comment by Jason-Dang** [comment]
I’ll get it done sometime tomorrow hopefully!

**Comment by Jason-Dang** [comment]
@Mac-Rae What conflicts? :joy:

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 5
           

Complexity increasing per file
==============================
- tests/unit/phpunit/data/BeanFactoryTest.php  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 5
           

Complexity increasing per file
==============================
- tests/unit/phpunit/data/BeanFactoryTest.php  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 5
           

Complexity increasing per file
==============================
- tests/unit/phpunit/data/BeanFactoryTest.php  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 5
           

Complexity increasing per file
==============================
- tests/unit/phpunit/data/BeanFactoryTest.php  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 3
           

Complexity increasing per file
==============================
- tests/unit/phpunit/data/BeanFactoryTest.php  1
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity increasing per file
==============================
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
- tests/unit/phpunit/data/BeanFactoryTest.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity increasing per file
==============================
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
- tests/unit/phpunit/data/BeanFactoryTest.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
- tests/unit/phpunit/data/BeanFactoryTest.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
- tests/unit/phpunit/data/BeanFactoryTest.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- tests/unit/phpunit/data/BeanFactoryTest.php  1
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

File: dev/ReplacerScript/BeanFactoryClassReplace.php [hunk 6]
```diff
@@ -0,0 +1,408 @@
+<?php
+
+$moduleClassList = [
```
Comment (Dillon-Brown) [inline dev/ReplacerScript/BeanFactoryClassReplace.php]: Proof that I actually code-reviewed every file :P.

**Comment by samus-aran** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- tests/SuiteCRM/Test/BeanFactoryTestCase.php  9
- tests/unit/phpunit/data/BeanFactoryTest.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/samus-aran/SuiteCRM/pullRequest?prid=4105113&bid=14113980)

File: dev/ReplacerScript/BeanFactoryClassReplace.php (hunk 6 above)
Comment (Jason-Dang) [reply dev/ReplacerScript/BeanFactoryClassReplace.php]: Good spot haha and thanks for removing the temp script :)

**Comment by Jason-Dang** [comment]
I'll take a look into what's going on with the unit/acceptance tests at the weekend and at the very least report my findings but hopefully I'll get things resolved.

**Comment by Mausino** [comment]
@Jason-Dang do you have links to PRs of stage 3 & 4 ?

**Comment by Jason-Dang** [comment]
@mausino No PRs for them yet I’m afraid.

**Comment by Mac-Rae** [comment]
Assessed :+1:

**Comment by Mac-Rae** [comment]
Finally :fireworks:

**Comment by Jason-Dang** [comment]
OMFG GG guys

**Pull request merged**